### ISSUE TITLE
Definition of regression loss function in YOLOv8

### ISSUE THREAD

**Issue by YQ0Sun** [ultralytics/ultralytics#8781]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, I would like to ask a question about the definition of regression loss function in YOLOv8, what kind of expressions are DFL Loss and CIoU Loss? Looking forward to your reply.

**Comment by glenn-jocher** [comment]
@YQ0Sun hi there! 👋

Great question! In YOLOv8, we use two primary types of loss functions for bounding box regression:

1. **Distribution Focal Loss (DFL)**: This is an improved version of the focal loss, designed to focus more on hard-to-classify examples. It works by adjusting the focal loss based on the distribution of classes, making the model more sensitive to minority classes.

2. **Complete Intersection over Union (CIoU) Loss**: CIoU Loss is an enhancement over the traditional IoU loss. It takes into account the overlap between predicted and ground truth boxes, the distance between the box centers, and the aspect ratio, providing a more comprehensive measure of the boxes' alignment.

Here's a simplified example of how CIoU loss might be calculated:

```python
def ciou_loss(preds, targets):
    # Calculate IoU, center distance, and aspect ratio terms
    iou = calculate_iou(preds, targets)
    center_distance = calculate_center_distance(preds, targets)
    aspect_ratio = calculate_aspect_ratio(preds, targets)
    
    # Combine terms for CIoU
    ciou = iou - (center_distance + aspect_ratio)
    return 1 - ciou  # Loss is 1 - CIoU
```

For DFL, the implementation details are more complex due to its focus on class distribution, and it's typically integrated directly into the model's architecture.

Hope this helps! Let us know if you have any more questions. 😊

**Comment by YQ0Sun** [comment]
> @YQ0Sun hi there! 👋
> 
> Great question! In YOLOv8, we use two primary types of loss functions for bounding box regression:
> 
> 1. **Distribution Focal Loss (DFL)**: This is an improved version of the focal loss, designed to focus more on hard-to-classify examples. It works by adjusting the focal loss based on the distribution of classes, making the model more sensitive to minority classes.
> 2. **Complete Intersection over Union (CIoU) Loss**: CIoU Loss is an enhancement over the traditional IoU loss. It takes into account the overlap between predicted and ground truth boxes, the distance between the box centers, and the aspect ratio, providing a more comprehensive measure of the boxes' alignment.
> 
> Here's a simplified example of how CIoU loss might be calculated:
> 
> ```python
> def ciou_loss(preds, targets):
>     # Calculate IoU, center distance, and aspect ratio terms
>     iou = calculate_iou(preds, targets)
>     center_distance = calculate_center_distance(preds, targets)
>     aspect_ratio = calculate_aspect_ratio(preds, targets)
>     
>     # Combine terms for CIoU
>     ciou = iou - (center_distance + aspect_ratio)
>     return 1 - ciou  # Loss is 1 - CIoU
> ```
> 
> For DFL, the implementation details are more complex due to its focus on class distribution, and it's typically integrated directly into the model's architecture.
> 
> Hope this helps! Let us know if you have any more questions. 😊

Thanks for the reply, I now know that bounding box regression is composed of two parts: the distribution focal loss and the complete intersection and concatenation loss, and I know from other sources that the formula for the loss function of bounding box regression is a weighted summation of these two parts of the loss function, is that true, please?
Loss = a Lciou + b Ldfl
where a and b are weighting factors.

**Comment by glenn-jocher** [comment]
Hey @YQ0Sun! 🌟

Absolutely, you've got it right! In YOLOv8, the total regression loss is indeed a weighted combination of the CIoU Loss (Lciou) and the Distribution Focal Loss (Ldfl). The formula you mentioned:

```python
Loss = a * Lciou + b * Ldfl
```

is spot on, where `a` and `b` are the weights that balance the contribution of each loss component. These weights help in fine-tuning the model's performance by emphasizing different aspects of the bounding box regression task. 

If you have any more questions or need further clarification, feel free to ask. Happy to help! 😊

**Comment by ARusDian** [comment]
may i ask about dfl loss,

if there's only a single class to detect, is dfl loss still important ?

**Comment by glenn-jocher** [comment]
DFL loss can still be beneficial for single-class detection as it helps refine bounding box predictions.

**Comment by alice984** [comment]
Hi @glenn-jocher, we need to replace the CIoU loss function in YOLOv8 with a new one, so can you give us the main steps for doing that, and the files where we have to add the new loss function?
We create a new class for the new loss function in "ultralytics/utils/new_loss.py then we add it to "ultralytics/utils/loss.py" and we add a message to verify if it has really changed, but it is still using CIoU.

**Comment by glenn-jocher** [comment]
Hi @alice984 — adding a new file won’t change YOLOv8’s regression loss unless you wire it into `BboxLoss`; replace the CIoU term inside `ultralytics/utils/loss.py` so your function is used, e.g.:

```python
# ultralytics/utils/loss.py
from ultralytics.utils.new_loss import custom_iou  # your IoU in [0, 1]

class BboxLoss(Module):
    def forward(self, pred_dist, pred_bboxes, anchor_points, target_bboxes, target_scores, target_scores_sum, fg_mask):
        weight = target_scores.sum(-1)[fg_mask].unsqueeze(-1)
        iou = custom_iou(pred_bboxes[fg_mask], target_bboxes[fg_mask])  # swap out CIoU
        loss_iou = ((1.0 - iou) * weight).sum() / target_scores_sum
        # DFL part unchanged...
        ...
        return loss_iou, loss_dfl
```

This replaces the call that currently uses `bbox_iou(..., CIoU=True)` in `BboxLoss.forward` as shown in the docs’ BboxLoss section; see the class and method in the [BboxLoss reference](https://docs.ultralytics.com/reference/utils/loss/#ultralyticsutilslossbboxloss). Loss weights are applied later in `v8DetectionLoss` (`loss[0] *= self.hyp.box`, `loss[2] *= self.hyp.dfl`); see [v8DetectionLoss in the loss reference](https://docs.ultralytics.com/reference/utils/loss/#ultralyticsutilslossv8detectionloss) and how the criterion is created in the model `loss()` flow in [tasks.py](https://docs.ultralytics.com/reference/nn/tasks/#loss).

If you’re not seeing the change, verify you’re running the edited install (not a pip wheel) by running `yolo checks` and confirming the package path, and restart your Python session/notebook after code edits. If you need further help, please open a new issue with your `yolo checks` output and the exact diff for `BboxLoss` for visibility.

**Comment by alice984** [comment]
Hi @glenn-jocher, thank you for your response. It was helpful,
Another question: now we are working with YOLOv8 on a radiographic dataset with low contrast and some noise, which makes it challenging to detect fine and elongated objects. We want to incorporate the Leung-Malik Filter Bank at the appropriate position in YOLOv8 to improve feature extraction. We want to see how to integrate them into a Conv2D layer, as an example, or as a parallel layer, and then add the filtered input to the next layer according to a specific criterion to reduce false negatives.
Here is the implementation: import torch
import torch.nn as nn
import numpy as np

def makeLMfilters():
    sup     = 49
    scalex  = np.sqrt(2) * np.array([1,2,3])
    norient = 6
    nrotinv = 12
    nbar  = len(scalex)*norient
    nedge = len(scalex)*norient
    nf    = nbar+nedge+nrotinv
    F     = np.zeros([sup,sup,nf])
    hsup  = (sup - 1)/2

    x = [np.arange(-hsup,hsup+1)]
    y = [np.arange(-hsup,hsup+1)]
    [x,y] = np.meshgrid(x,y)
    orgpts = np.array([x.flatten(), y.flatten()])

    def gaussian1d(sigma, mean, x, ord):
        x_ = x - mean
        var = sigma**2
        g1 = (1/np.sqrt(2*np.pi*var)) * np.exp(-x_**2/(2*var))
        if ord == 0:
            return g1
        elif ord == 1:
            return -g1 * (x_/var)
        else:
            return g1 * ((x_**2 - var) / (var**2))

    def makefilter(scale, phasex, phasey, pts, sup):
        gx = gaussian1d(3*scale, 0, pts[0,...], phasex)
        gy = gaussian1d(scale, 0, pts[1,...], phasey)
        return np.reshape(gx*gy, (sup, sup))

    count = 0
    for scale in range(len(scalex)):
        for orient in range(norient):
            angle = (np.pi * orient)/norient
            c, s = np.cos(angle), np.sin(angle)
            rotpts = np.dot(np.array([[c, -s],[s, c]]), orgpts)
            F[:,:,count] = makefilter(scalex[scale], 0, 1, rotpts, sup)
            F[:,:,count+nedge] = makefilter(scalex[scale], 0, 2, rotpts, sup)
            count += 1
    count = nbar+nedge
    scales = np.sqrt(2) * np.array([1,2,3,4])
    for i in range(len(scales)):
        var = scales[i]**2
        x,y = np.ogrid[-hsup:hsup+1,-hsup:hsup+1]
        g = (1/np.sqrt(2*np.pi*var))*np.exp(-(x*x+y*y)/(2*var))
        F[:,:,count] = g; count += 1
        h = g*((x*x+y*y)-var)/(var**2)
        F[:,:,count] = h; count += 1
        F[:,:,count] = h; count += 1
    return F.astype(np.float32)

**Comment by glenn-jocher** [comment]
Nice idea for this kind of low‑contrast data. You can treat the Leung–Malik bank as a frozen `Conv2d` and plug it in near the input of YOLOv8.

A minimal way to wrap your filters as a convolution is:

```python
import torch
import torch.nn as nn
import numpy as np

class LMFilterBank(nn.Module):
    def __init__(self):
        super().__init__()
        F = makeLMfilters()              # (sup, sup, nf)
        sup, _, nf = F.shape
        kernels = np.moveaxis(F, -1, 0)  # (nf, sup, sup)

        self.conv = nn.Conv2d(
            in_channels=1,               # radiographs often 1‑channel
            out_channels=nf,
            kernel_size=sup,
            padding=sup // 2,
            bias=False,
        )
        with torch.no_grad():
            self.conv.weight.copy_(torch.from_numpy(kernels)[:, None, ...])
        self.conv.weight.requires_grad_(False)

    def forward(self, x):
        return self.conv(x)
```

Integration options in YOLOv8 are then, for example:

1) Replace or precede the first backbone layer: run `LMFilterBank` on the input image and feed its output (or a 1×1‑conv projection of it) into the first YOLO layer instead of the raw image.

2) Parallel branch: apply `LMFilterBank` to the input (or an early feature map), concatenate its response with the standard features along the channel dimension, and use a small learnable conv (e.g. 1×1 or 3×3) to fuse them before passing to the next YOLO block; inside that fusion block you can implement your “criterion” (for example, a learned gate using `sigmoid` on the LM response and multiplying the backbone features).

The exact wiring (which layer to tap, how many channels to fuse back to, and what gating rule to use) is a research choice, so I’d suggest starting with a simple concat‑and‑1×1‑conv at the very first or second backbone stage, measure the impact on false negatives, and then iterate. If you run into a specific integration error (shape mismatch, YAML change, etc.), please share the error text and a small code/YAML snippet in a new issue so we can take a closer look.
